### PR TITLE
Import districts

### PR DESCRIPTION
**Pull request by illusionOfParadise**
https://app.clubhouse.io/enmarche/story/4363/import-des-circonscriptions

### CODE REVIEW

File: app/migrations/Version20180608173623.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+final class Version20180608173623 extends AbstractMigration
+{
+    public function up(Schema $schema): void
+    {
+        $this->addSql('CREATE TABLE districts (
+                            id INT UNSIGNED AUTO_INCREMENT NOT NULL, 
+                            adherent_id INT UNSIGNED DEFAULT NULL, 
+                            countries LONGTEXT NOT NULL COMMENT \'(DC2Type:simple_array)\', 
+                            code VARCHAR(6) NOT NULL, 
+                            number SMALLINT UNSIGNED NOT NULL,
```
Comment (hhamon) [inline app/migrations/Version20180608173623.php]: `number` should be safen with backticks.

File: src/Command/ImportDistrictsCommand.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    {
+        $this->io = new SymfonyStyle($input, $output);
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $file = $input->getArgument('file');
+        $districtsFile = $input->getArgument('districtsFile');
+        $countriesFile = $input->getArgument('countriesFile');
+        if (!is_file($file)) {
```
Comment (hhamon) [inline src/Command/ImportDistrictsCommand.php]:
```
if (!is_readable($file = $input->getArgument('file'))) {
    // ...
}
```

File: src/Command/ImportDistrictsCommand.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            $this->io->error("$file is not a file");
+
+            return 1;
+        }
+        if (!is_file($districtsFile)) {
```
Comment (hhamon) [inline src/Command/ImportDistrictsCommand.php]:
```
if (!is_readable($file = $input->getArgument('districtsFile'))) {
    // ...
}
```

File: src/Command/ImportDistrictsCommand.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            $this->io->error("$districtsFile is not a file");
+
+            return 1;
+        }
+        if (!is_file($countriesFile)) {
```
Comment (hhamon) [inline src/Command/ImportDistrictsCommand.php]:
```
if (!is_readable($file = $input->getArgument('countriesFile'))) {
    // ...
}
```

File: src/Deputy/DistrictLoader.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            $geoShape = $this->geometryFactory->mergeGeoJsonGeometries($geoCountries);
+        }
+
+        if ($existingDistrict = $this->em->getRepository(District::class)->findOneBy(['code' => $district['circo_ID']])) {
+            return $existingDistrict->update(
+                $countries,
+                $district['nom_circo'],
+                $geoShape
+            );
+        } else {
```
Comment (hhamon) [inline src/Deputy/DistrictLoader.php]: `else` statement is not needed.

File: src/Entity/Adherent.php [hunk 6]
```diff
@@ -1106,6 +1113,16 @@ public function removeEmptyCoordinatorManagedAreas(): void
         }
     }
 
+    public function getManagedDistrict(): ?District
+    {
+        return $this->managedDistrict;
+    }
+
+    public function setManagedDistrict(District $district): void
```
Comment (hhamon) [inline src/Entity/Adherent.php]: `?District` as it's nullable.

File: src/Entity/District.php [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+     * @var int|null
+     *
+     * @ORM\Id
+     * @ORM\Column(type="integer", options={"unsigned": true})
+     * @ORM\GeneratedValue
+     */
+    private $id;
+
+    /**
+     * The the managed district countries.
```
Comment (hhamon) [inline src/Entity/District.php]: ` The` is written twice ^^

File: src/Entity/District.php [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+     * @var array
+     *
+     * @ORM\Column(type="simple_array")
+     */
+    private $countries = [];
+
+    /**
+     * @var string
+     *
+     * @ORM\Column(length=6, unique=true)
```
Comment (hhamon) [inline src/Entity/District.php]: Better to define the unique constraint in the `@Table` annotation

File: src/Entity/District.php [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+    /**
+     * @var string
+     *
+     * @ORM\Column(length=6)
+     */
+    private $code;
+
+    /**
+     * @var string
```
Comment (ottaviano) [inline src/Entity/District.php]: int

File: src/Entity/District.php [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+    /**
+     * @var Polygon|MultiPolygon
+     *
+     * @ORM\Column(type="geometry")
+     */
+    private $geoShape;
+
+    /**
+     * @ORM\OneToOne(targetEntity="AppBundle\Entity\Adherent", inversedBy="managedDistrict")
+     * @ORM\JoinColumn(name="adherent_id", referencedColumnName="id")
```
Comment (ottaviano) [inline src/Entity/District.php]: this is the default value for join the column name

File: src/Entity/District.php [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+        return $this->geoShape->toArray();
+    }
+
+    public function setAdherent(Adherent $adherent = null)
+    {
+        $this->adherent = $adherent;
+    }
+
+    public function update(
+        array $countries,
```
Comment (ottaviano) [inline src/Entity/District.php]: CS: can you inline the args pls?

File: src/Geo/GeometryFactory.php [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+namespace AppBundle\Geo;
+
+use CrEOF\Spatial\PHP\Types\Geometry\GeometryInterface;
+use CrEOF\Spatial\PHP\Types\Geometry\MultiPolygon;
+use CrEOF\Spatial\PHP\Types\Geometry\Point;
+use CrEOF\Spatial\PHP\Types\Geometry\Polygon;
+
+class GeometryFactory
+{
+    public function createPointFromGeoPoint(string $geoPoint): Point
```
Comment (ottaviano) [inline src/Geo/GeometryFactory.php]: `createPointFromString` will be clearer, no?

File: src/Geo/GeometryFactory.php [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        $geoPoint = explode(', ', $geoPoint);
+
+        return new Point($geoPoint[1], $geoPoint[0]);
```
Comment (ottaviano) [inline src/Geo/GeometryFactory.php]: `new Point($geoPoint[0], $geoPoint[1]);` no? (`[0]` first, `[1]` after)

File: src/Deputy/DistrictLoader.php [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+            $countries = explode(',', str_replace(' ', '', $district['code_pays']));
+            $geoCountries = array_filter($this->geoCountries, function ($country) use ($countries) {
+                return in_array($country['properties']['code'], $countries);
+            });
+
+            if (0 === count($geoCountries)) {
+                throw new \RuntimeException("Countries GeoJSON file doesn't contain countries with codes '$countries'");
+            }
+            $geoShape = $this->geometryFactory->mergeGeoJsonGeometries($geoCountries);
```
Comment (ottaviano) [inline src/Deputy/DistrictLoader.php]: you make your own multipolygon by merging the polygons of each country?
maybe you can use this [official district geo data](https://public.opendatasoft.com/explore/dataset/circonscriptions-legislatives-2017/export/?dataChart=eyJxdWVyaWVzIjpbeyJjb25maWciOnsiZGF0YXNldCI6ImNpcmNvbnNjcmlwdGlvbnMtbGVnaXNsYXRpdmVzLTIwMTciLCJvcHRpb25zIjp7fX0sImNoYXJ0cyI6W3sidHlwZSI6ImNvbHVtbiIsImZ1bmMiOiJDT1VOVCIsInNjaWVudGlmaWNEaXNwbGF5Ijp0cnVlLCJjb2xvciI6IiMyQzNGNTYifV0sInhBeGlzIjoicmVmIiwibWF4cG9pbnRzIjo1MCwic29ydCI6IiJ9XSwidGltZXNjYWxlIjoiIiwiZGlzcGxheUxlZ2VuZCI6dHJ1ZX0%3D) file (but only for France)

File: app/config/config.yml [hunk 15]
```diff
@@ -104,6 +104,10 @@ doctrine:
             uuid:           Ramsey\Uuid\Doctrine\UuidType
             phone_number:   Misd\PhoneNumberBundle\Doctrine\DBAL\Types\PhoneNumberType
             geo_point:      AppBundle\Doctrine\DBAL\Types\GeoPointType
+            geometry:       CrEOF\Spatial\DBAL\Types\GeometryType
+            point:          CrEOF\Spatial\DBAL\Types\Geometry\PointType
```
[the remark this answers is not in the document]
Comment (illusionOfParadise) [reply app/config/config.yml]: Actually only `point` is not used. I can remove it.

File: src/Entity/District.php [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+    /**
+     * @var Polygon|MultiPolygon
+     *
+     * @ORM\Column(type="geometry")
+     */
+    private $geoShape;
+
+    /**
+     * @ORM\OneToOne(targetEntity="AppBundle\Entity\Adherent", inversedBy="managedDistrict")
+     * @ORM\JoinColumn
```
Comment (ottaviano) [inline src/Entity/District.php]: it's not needed, you can remove JoinColumn.

File: src/Entity/District.php [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+    private $adherent;
+
+    /**
+     * @throws \InvalidArgumentException if $geoShape is invalid
+     */
+    public function __construct(
+        array $countries,
+        string $name,
+        string $code,
+        string $number,
```
Comment (ottaviano) [inline src/Entity/District.php]: int

File: src/Entity/District.php [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+    public const FRANCE = 'FR';
+
+    /**
+     * @var int|null
+     *
+     * @ORM\Id
+     * @ORM\Column(type="integer", options={"unsigned": true})
+     * @ORM\GeneratedValue
+     */
+    private $id;
```
Comment (Remg) [inline src/Entity/District.php]: Should we define a `getId` method ?

File: src/Deputy/DistrictLoader.php [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+
+        $i = 0;
+
+        foreach ($districts as $district) {
+            $this->createOrUpdateDistrict($district);
+            $this->em->persist($this->createOrUpdateDistrict($district));
+            $this->em->flush();
+
+            if (0 === ++$i % 1000 || $i === count($districts)) {
+                $this->em->clear();
```
Comment (B-Galati) [inline src/Deputy/DistrictLoader.php]: `flush()` should be in that condition too no ?

Comment (illusionOfParadise) [reply src/Deputy/DistrictLoader.php]: Not really, I want to flush after each insert/update.

**Pull request merged**